diff --git a/fs_mgr/fs_mgr.cpp b/fs_mgr/fs_mgr.cpp
index 560b1ab..dc57b35 100644
--- a/fs_mgr/fs_mgr.cpp
+++ b/fs_mgr/fs_mgr.cpp
@@ -170,6 +170,10 @@ static bool should_force_check(int fs_stat) {
             FS_STAT_SET_RESERVED_BLOCKS_FAILED | FS_STAT_ENABLE_ENCRYPTION_FAILED);
 }
 
+static bool should_set_no_kernel_check() {
+    return android::base::GetBoolProperty("ro.fsck_no_kernel_check", false);
+}
+
 static void check_fs(const std::string& blk_device, const std::string& fs_type,
                      const std::string& target, int* fs_stat) {
     int status;
@@ -259,22 +263,17 @@ static void check_fs(const std::string& blk_device, const std::string& fs_type,
             }
         }
     } else if (is_f2fs(fs_type)) {
-        const char* f2fs_fsck_argv[] = {F2FS_FSCK_BIN,     "-a", "-c", "10000", "--debug-cache",
-                                        blk_device.c_str()};
-        const char* f2fs_fsck_forced_argv[] = {
-                F2FS_FSCK_BIN, "-f", "-c", "10000", "--debug-cache", blk_device.c_str()};
-
-        if (should_force_check(*fs_stat)) {
-            LINFO << "Running " << F2FS_FSCK_BIN << " -f -c 10000 --debug-cache "
-                  << realpath(blk_device);
-            ret = logwrap_fork_execvp(ARRAY_SIZE(f2fs_fsck_forced_argv), f2fs_fsck_forced_argv,
-                                      &status, false, LOG_KLOG | LOG_FILE, false, FSCK_LOG_FILE);
-        } else {
-            LINFO << "Running " << F2FS_FSCK_BIN << " -a -c 10000 --debug-cache "
-                  << realpath(blk_device);
-            ret = logwrap_fork_execvp(ARRAY_SIZE(f2fs_fsck_argv), f2fs_fsck_argv, &status, false,
-                                      LOG_KLOG | LOG_FILE, false, FSCK_LOG_FILE);
+        const char* force = should_force_check(*fs_stat) ? "-f" : "-a";
+        std::vector<const char*> f2fs_fsck_argv = {F2FS_FSCK_BIN, force, "-c", "10000",
+                                                   "--debug-cache"};
+        if (should_set_no_kernel_check()) {
+            f2fs_fsck_argv.push_back("--no-kernel-check");
         }
+        f2fs_fsck_argv.push_back(blk_device.c_str());
+
+        LINFO << "Running " << android::base::Join(f2fs_fsck_argv, " ");
+        ret = logwrap_fork_execvp(f2fs_fsck_argv.size(), f2fs_fsck_argv.data(), &status, false,
+                                  LOG_KLOG | LOG_FILE, false, FSCK_LOG_FILE);
         if (ret < 0) {
             /* No need to check for error in fork, we can't really handle it now */
             LERROR << "Failed trying to run " << F2FS_FSCK_BIN;
